hacks/ifs.man
.TH XScreenSaver 1 "" "X Version 11"
.SH NAME
ifs \- draws spinning, colliding iterated-function-system images
.SH SYNOPSIS
.B ifs
[\-\-display \fIhost:display.screen\fP]
[\-\-visual \fIvisual\fP]
[\-\-window]
[\-\-root]
[\-\-window\-id \fInumber\fP]
[\-\-no\-db]
[\-\-delay \fInumber\fP]
[\-\-detail \fInumber\fP]
[\-\-colors \fInumber\fP]
[\-\-functions \fInumber\fP]
[\-\-iterate | \-\-recurse]
[\-\-no\-rotate]
[\-\-no\-scale]
[\-\-no\-translate]
[\-\-fps]
.SH DESCRIPTION
The \fIifs\fP program draws spinning, colliding iterated-function-system images.
.SH OPTIONS
.TP 8
.B \-\-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-\-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-\-root
Draw on the root window.
.TP 8
.B \-\-window\-id \fInumber\fP
Draw on the specified window.
.TP 8
.B \-\-no\-db
Disable double-buffering.
.TP 8
.B \-\-delay \fInumber\fP
Per-frame delay, in microseconds.  Default: 20000
.TP 8
.B \-\-colors \fInumber\fP
Number of colours to use. Default: 200
.TP 8
.B \-\-detail \fInumber\fP
In \fI\-\-iterate\fP mode, number of times to randomly iterate the
functions, in thousands.  In \fI\-\-recurse\fP mode, number of times
to apply functions (recursion depth) before drawing each point.
Default: 9
.TP 8
.B \-\-functions \fInumber\fP
Number of functions to be iterated. Default: 3
.TP 8
.B \-\-iterate
Calculate by iteratively applying the functions in a random order,
usually faster than \fI\-\-recurse\fP.  This is the default.
.TP 8
.B \-\-recurse
Calculate by recursively applying all combinations of the functions.
This is the historical behavior and may produce neater output than
\fI\-\-iterate\fP.
.TP 8
.B \-\-no-rotate
Disable the rotation component of the functions.
.TP 8
.B \-\-no-scale
Disable the scaling component of the functions.
.TP 8
.B \-\-no-translate
Disable the varying translation component of the functions.
.TP 8
.B \-\-no-multi
Turn off multi-coloured mode, only one colour is used to colour the whole set.
.TP 8
.B \-\-fps
Display the current frame rate and CPU load.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.TP 8
.B XSCREENSAVER_WINDOW
The window ID to use with \fI\-\-root\fP.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1),
.BR ifs (MANSUFFIX)
.SH COPYRIGHT
Copyright \(co Chris Le Sueur and Robby Griffin, 2005-2006

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.
.SH AUTHOR
This version by Chris Le Sueur <thefishface@gmail.com>, Feb 2005
Many improvements by Robby Griffin <rmg@terc.edu>, Mar 2006


po/ChangeLog
2002-11-25  Kostas Papadimas <pkst@gmx.net>

	* el.po: Added Greek translation

2002-11-22  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Updated Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-17  Dmitry G. Mastrukov  <dmitry@taurussoft.org>

	* be.po: Added Belarusian translation
	* from Belarusian team <i18n@infonet.by>.

2002-11-06  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-24  Christian Meyer  <chrisime@gnome.org>

	* de.po: Updated German translation.

2002-09-23  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-22  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-20  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-19  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Updated Vietnamese file

2002-09-17  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Fixed the (C) issue in the english message.

2002-09-16  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish translation.

2002-09-08  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.
			This should fix the "error" on the statuspages.

2002-09-07  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-05  Christian Neumair  <chris@gnome-de.org>

	* de.po: Updated German translation.

2002-09-03  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmål) translation.

2002-08-29  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-08-29  Tõivo Leedjärv  <toivo@linux.ee>

	* et.po: Fixed Estonian translation.

2002-08-29  Christian Neumair  <christian-neumair@web.de>

	* de.po: Updated German translation.

2002-08-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Partial update of Danish translation.

2002-08-23  Andras Timar  <timar@gnome.huy

	* hu.po: Added Hungarian translation (partial).

2002-08-13  Christian Neumair  <christian-neumair@web.de>

	* de.po: Started German translation.

2002-08-11  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Dutch translation updated by Tino Meinen.

2002-07-28  Christophe Merlet <christophe@merlet.net>

	* fr.po: Updated French translation from contribution of
	Sun G11n <gnome_int_l10n@ireland.sun.com>.

2002-07-22  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-07-12  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po,wa.po: Updated Vietnamese and Walloon files

2002-06-26  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po: Fixed a duplicated string.

2002-06-24  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-25  Christian Rose  <menthos@menthos.com>

	* fr.po, pl.po, ru.po, uk.po, wa.po: Added from xscreensaver module.

2002-06-21  Changwoo Ryu  <cwryu@debian.org>

	* ko.po: Added Korean translation from Young-Ho Cha
	<ganadist@chollian.net>.

2002-06-19  Christian Rose  <menthos@menthos.com>

	* sv.po: Updated Swedish translation somewhat.

2002-06-18  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO, xscreensaver.pot: Updated from 4.0.5 release.
	
2002-06-16  Carlos Perello Marin <carlos@gnome-db.org>

	* INFO: Added extra info. I will take all translations from here, you
	don't need to send me it by mail.

2002-06-16  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* vi.po: Added Vietnamese file

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-14  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.

2002-06-12  Abel Cheung  <maddog@linux.org.hk>

	* zh_TW.po: New traditional Chinese translation.

2002-06-09  Takeshi Aihana <aihana@gnome.gr.jp>

	* ja.po: Added as Japanese translation.

2002-06-06  Christian Rose  <menthos@menthos.com>

	* sv.po: Added from xscreensaver module.

2002-06-06  Ole Laursen  <olau@hardworking.dk>

	* da.po: Added and halfway-updated from xscreensaver module.

2002-06-06  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

	* et.po: Added from xscreensaver module.

2002-06-06  Carlos Perello Marin <carlos@gnome-db.org>

	* es.po,nl.po,no.po,pt.po,sk.po,zh_CN.po: Added from xscreensaver
	module.
	* xscreensaver.pot: Added from latest release (4.04.2).
	* INFO: Added info about this module.

2002-06-04  Vincent van Adrighem <V.vanAdrighem@dirck.mine.nu>

        * nl.po: Added Dutch translation.

2002-05-26  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-25  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Pablo Gonzalo del Campo

	* es.po: Updated Spanish Translation.

2002-05-23  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Backported Simplified Chinese translation from HEAD.

2002-05-19  Pablo Gonzalo del Campo <pablodc@bigfoot.com>

	* es.po: Updated Spanish Translation.
						
2002-05-17  Wang Jian  <lark@linux.net.cn>

	* zh_CN.po: Added Simplified Chinese translation by
	He Qiangqiang <carton@linux.net.cn>.

2002-05-11  Stanislav Visnovsky  <visnovsky@nenya.ms.mff.cuni.cz>

        * sk.po: Backported Slovak translation from HEAD.
	
2002-05-01  GermÃ¡n Poo-CaamaÃ±o  <gpoo@ubiobio.cl>

	* es.po: Updated Spanish translation from
	Guillermo Movia <gmovia@hotpop.com> and converted to UTF-8

2002-04-29  Pablo Saratxaga  <pablo@mandrakesoft.com>

	* wa.po: Updated Walloon file

2002-04-27  Duarte Loreto <happyguy_pt@hotmail.com>

        * pt.po: Added Portuguese translation.

2002-04-26  TÃµivo LeedjÃ€rv  <toivo@linux.ee>

        * et.po: Updated Estonian translation.

2002-04-26  Ole Laursen  <olau@hardworking.dk>

	* da.po: Updated Danish translation.

2002-04-21  Kjartan Maraas  <kmaraas@gnome.org>

	* no.po: Updated Norwegian (bokmÃ¥l) translation.


hacks/ifs.c
/* Copyright © Chris Le Sueur and Robby Griffin, 2005-2006

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE X CONSORTIUM BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

Ultimate thanks go to Massimino Pascal, who created the original
xscreensaver hack, and inspired me with it's swirly goodness. This
version adds things like variable quality, number of functions and also
a groovier colouring mode.

This version by Chris Le Sueur <thefishface@gmail.com>, Feb 2005
Many improvements by Robby Griffin <rmg@terc.edu>, Mar 2006
Multi-coloured mode added by Jack Grahl <j.grahl@ucl.ac.uk>, Jan 2007
*/

#include <assert.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "screenhack.h"

typedef struct {
  float r, s, tx, ty;   /* Rotation, Scale, Translation X & Y */
  float ro, rt, rc;     /* Old Rotation, Rotation Target, Rotation Counter */
  float so, st, sc;     /* Old Scale, Scale Target, Scale Counter */
  float sa, txa, tya;   /* Scale change, Translation change */

  int ua, ub, utx;      /* Precomputed combined r,s,t values */
  int uc, ud, uty;      /* Precomputed combined r,s,t values */

} Lens;	

struct state {
  Display *dpy;
  Window window;
  GC gc;
  Drawable backbuffer;
  XColor *colours;
  int ncolours;
  int ccolour;
  int blackColor, whiteColor;

  int width, widthb, height;
  int width8, height8;
  unsigned int *board;
  XRectangle pointbuf[1000];
  int npoints;
  int xmin, xmax, ymin, ymax;
  int x, y;
  int pscale;

  int delay;

  int lensnum;
  Lens *lenses;
  int length;
  int mode;
  Bool recurse;
  Bool multi;
  Bool translate, scale, rotate;
};

#define getdot(x,y) (st->board[((y)*st->widthb)+((x)>>5)] &  (1<<((x) & 31)))
#define setdot(x,y) (st->board[((y)*st->widthb)+((x)>>5)] |= (1<<((x) & 31)))

static double
myrandom(float up)
{
  return (((double)random() / RAND_MAX) * up);
}

static const char *ifs_defaults [] = {
/*  ".lowrez:             true", */
  ".background:		Black",
  "*lensnum:		3",
  "*fpsSolid:		true",
  "*length:		9",
  "*mode:		0",
  "*colors:		200",
  "*delay:		20000",
  "*translate:		True",
  "*ifsScale:		True",
  "*rotate:		True",
  "*recurse:		False",
  "*multi:              True",
# ifdef HAVE_JWXYZ	/* Don't second-guess Quartz's double-buffering */
  "*doubleBuffer:	False",
#else
  "*doubleBuffer:	True",
#endif
#ifdef HAVE_MOBILE
  "*ignoreRotation:     True",
#endif
  0
};

static XrmOptionDescRec ifs_options [] = {
  { "-detail",		".length",	XrmoptionSepArg, 0 },
  { "-delay",		".delay",	XrmoptionSepArg, 0 },
  { "-mode",		".mode",	XrmoptionSepArg, 0 },
  { "-colors",		".colors",	XrmoptionSepArg, 0 },
  { "-functions",	".lensnum", 	XrmoptionSepArg, 0 },
  { "-no-translate",	".translate",   XrmoptionNoArg, "False" },
  { "-no-scale",	".ifsScale",	XrmoptionNoArg, "False" },
  { "-no-rotate",	".rotate",	XrmoptionNoArg, "False" },
  { "-recurse",		".recurse",	XrmoptionNoArg, "True" },
  { "-iterate",		".recurse",	XrmoptionNoArg, "False" },
  { "-multi",           ".multi",       XrmoptionNoArg, "True" },
  { "-no-multi",        ".multi",       XrmoptionNoArg, "False" },
  { "-db",		".doubleBuffer",XrmoptionNoArg, "True" },
  { "-no-db",		".doubleBuffer",XrmoptionNoArg, "False" },
  { 0, 0, 0, 0 }
};


/* Draw all the queued points on the backbuffer */
static void
drawpoints(struct state *st)
{
  XFillRectangles(st->dpy, st->backbuffer, st->gc, st->pointbuf, st->npoints);
  st->npoints = 0;
}

/* Set a point to be drawn, if it hasn't been already.
 * Expects coordinates in 256ths of a pixel. */
static void
sp(struct state *st, int x, int y)
{
  if (x < 0 || x >= st->width8 || y < 0 || y >= st->height8)
    return;

  x >>= 8;
  y >>= 8;

  if (getdot(x, y)) return;
  setdot(x, y);

  if (x < st->xmin) st->xmin = x;
  if (x > st->xmax) st->xmax = x;
  if (y < st->ymin) st->ymin = y;
  if (y > st->ymax) st->ymax = y;

  st->pointbuf[st->npoints].x = x;
  st->pointbuf[st->npoints].y = y;
  st->pointbuf[st->npoints].width =
    st->pointbuf[st->npoints].height = st->pscale;
  st->npoints++;

  if (st->npoints >= countof(st->pointbuf)) {
    drawpoints(st);
  }
}


/* Precompute integer values for matrix multiplication and vector
 * addition. The matrix multiplication will go like this (see iterate()):
 *   |x2|     |ua ub|   |x|     |utx|
 *   |  |  =  |     | * | |  +  |   |
 *   |y2|     |uc ud|   |y|     |uty|
 * 
 * There is an extra factor of 2^10 in these values, and an extra factor of
 * 2^8 in the coordinates, in order to implement fixed-point arithmetic.
 */
static void
lensmatrix(struct state *st, Lens *l)
{
  l->ua = 1024.0 * l->s * cos(l->r);
  l->ub = -1024.0 * l->s * sin(l->r);
  l->uc = -l->ub;
  l->ud = l->ua;
  l->utx = 131072.0 * st->width * (l->s * (sin(l->r) - cos(l->r))
				   + l->tx / 16 + 1);
  l->uty = -131072.0 * st->height * (l->s * (sin(l->r) + cos(l->r))
				     + l->ty / 16 - 1);
}

static void
CreateLens(struct state *st,
           float nr,
           float ns,
           float nx,
           float ny,
           Lens *newlens)
{
  newlens->sa = newlens->txa = newlens->tya = 0;
  if (st->rotate) {
    newlens->r = newlens->ro = newlens->rt = nr;
    newlens->rc = 1;
  }
  else newlens->r = 0;

  if (st->scale) {
    newlens->s = newlens->so = newlens->st = ns;
    newlens->sc = 1;
  }
  else newlens->s = 0.5;

  newlens->tx = nx;
  newlens->ty = ny;

  lensmatrix(st, newlens);
}
	
static void
mutate(struct state *st, Lens *l)
{
  if (st->rotate) {
    float factor;
    if(l->rc >= 1) {
      l->rc = 0;
      l->ro = l->rt;
      l->rt = myrandom(4) - 2;
    }
    factor = (sin((-M_PI / 2.0) + M_PI * l->rc) + 1.0) / 2.0;
    l->r = l->ro + (l->rt - l->ro) * factor;
    l->rc += 0.01;
  }
  if (st->scale) {
    float factor;
    if (l->sc >= 1) {
      /* Reset counter, obtain new target value */
      l->sc = 0;
      l->so = l->st;
      l->st = myrandom(2) - 1;
    }
    factor = (sin((-M_PI / 2.0) + M_PI * l->sc) + 1.0) / 2.0;
    /* Take average of old target and new target, using factor to *
     * weight. It's computed sinusoidally, resulting in smooth,   *
     * rhythmic transitions.                                      */
    l->s = l->so + (l->st - l->so) * factor;
    l->sc += 0.01;
  }
  if (st->translate) {
    l->txa += myrandom(0.004) - 0.002;
    l->tya += myrandom(0.004) - 0.002;
    l->tx += l->txa;
    l->ty += l->tya;
    if (l->tx > 6) l->txa -= 0.004;
    if (l->ty > 6) l->tya -= 0.004;
    if (l->tx < -6) l->txa += 0.004;
    if (l->ty < -6) l->tya += 0.004;
    if (l->txa > 0.05 || l->txa < -0.05) l->txa /= 1.7;
    if (l->tya > 0.05 || l->tya < -0.05) l->tya /= 1.7;
  }
  if (st->rotate || st->scale || st->translate) {
    lensmatrix(st, l);
  }
}


#define STEPX(l,x,y) (((l)->ua * (x) + (l)->ub * (y) + (l)->utx) >> 10)
#define STEPY(l,x,y) (((l)->uc * (x) + (l)->ud * (y) + (l)->uty) >> 10)
/*#define STEPY(l,x,y) (((l)->ua * (y) - (l)->ub * (x) + (l)->uty) >> 10)*/

/* Calls itself <lensnum> times - with results from each lens/function.  *
 * After <length> calls to itself, it stops iterating and draws a point. */
static void
recurse(struct state *st, int x, int y, int length, int p)
{
  int i;
  Lens *l;

  if (length == 0) {
    if (p == 0) 
      sp(st, x, y);
    else {
      l = &st->lenses[p];
      sp(st, STEPX(l, x, y), STEPY(l, x, y));
    }
  }
  else {
    for (i = 0; i < st->lensnum; i++) {
      l = &st->lenses[i];
      recurse(st, STEPX(l, x, y), STEPY(l, x, y), length - 1, p);
    }
  }
}

/* Performs <count> random lens transformations, drawing a point at each
 * iteration after the first 10.
 */
static void
iterate(struct state *st, int count, int p)
{
  int i;
  Lens *l;
  int x = st->x;
  int y = st->y;
  int tx;

# define STEP()                              \
    l = &st->lenses[random() % st->lensnum]; \
    tx = STEPX(l, x, y);                     \
    y = STEPY(l, x, y);                      \
    x = tx

  for (i = 0; i < 10; i++) {
    STEP();
  }

  for ( ; i < count; i++) {
    STEP();
    if (p == 0)
      sp(st, x, y);
    else
      {
	l = &st->lenses[p];
	sp(st, STEPX(l, x, y), STEPY(l, x, y));
      }
  }

# undef STEP

  st->x = x;
  st->y = y;
}

/* Come on and iterate, iterate, iterate and sing... *
 * Yeah, this function just calls iterate, mutate,   *
 * and then draws everything.                        */
static unsigned long
ifs_draw (Display *dpy, Window window, void *closure)
{
  struct state *st = (struct state *) closure;
  int i;
  int xmin = st->xmin, xmax = st->xmax, ymin = st->ymin, ymax = st->ymax;
  int partcolor, x, y;
  

  /* erase whatever was drawn in the previous frame */
  if (xmin <= xmax && ymin <= ymax) {
    XSetForeground(st->dpy, st->gc, st->blackColor);
    XFillRectangle(st->dpy, st->backbuffer, st->gc,
		   xmin, ymin,
		   xmax - xmin + 1 * st->pscale,
                   ymax - ymin + 1 * st->pscale);
    st->xmin = st->width  + 1;
    st->xmax = st->ymax =  -1;
    st->ymin = st->height + 1;
  }

  st->ccolour++;
  st->ccolour %= st->ncolours;

  /* calculate and draw points for this frame */
  x = st->width << 7;
  y = st->height << 7;
  
  if (st->multi) {
    for (i = 0; i < st->lensnum; i++) {  
      partcolor = st->ccolour * (i+1);
      partcolor %= st->ncolours;
      XSetForeground(st->dpy, st->gc, st->colours[partcolor].pixel);
      memset(st->board, 0, st->widthb * st->height * sizeof(*st->board));
      if (st->recurse)   
	recurse(st, x, y, st->length - 1, i);
      else
	iterate(st, pow(st->lensnum, st->length - 1), i);
      if (st->npoints) 
	drawpoints(st);
    }
  } 
  else {
    
    XSetForeground(st->dpy, st->gc, st->colours[st->ccolour].pixel);
    memset(st->board, 0, st->widthb * st->height * sizeof(*st->board));
    if (st->recurse)
      recurse(st, x, y, st->length, 0);
    else
      iterate(st, pow(st->lensnum, st->length), 0);
    if (st->npoints)
      drawpoints(st);
  }
  
  /* if we just drew into a buffer, copy the changed area (including
   * erased area) to screen */
  if (st->backbuffer != st->window
      && ((st->xmin <= st->xmax && st->ymin <= st->ymax)
	  || (xmin <= xmax && ymin <= ymax))) {
    if (st->xmin < xmin) xmin = st->xmin;
    if (st->xmax > xmax) xmax = st->xmax;
    if (st->ymin < ymin) ymin = st->ymin;
    if (st->ymax > ymax) ymax = st->ymax;
    XCopyArea(st->dpy, st->backbuffer, st->window, st->gc,
	      xmin, ymin,
	      xmax - xmin + 1, ymax - ymin + 1,
	      xmin, ymin);
  }

  for(i = 0; i < st->lensnum; i++) {
    mutate(st, &st->lenses[i]);
  }

  return st->delay;
}

static void
ifs_reshape (Display *, Window, void *, unsigned int, unsigned int);

static void *
ifs_init (Display *d_arg, Window w_arg)
{
  struct state *st = (struct state *) calloc (1, sizeof(*st));
  int i;
  XWindowAttributes xgwa;
	
  /* Initialise all this X shizzle */
  st->dpy = d_arg;
  st->window = w_arg;

  st->blackColor = BlackPixel(st->dpy, DefaultScreen(st->dpy));
  st->whiteColor = WhitePixel(st->dpy, DefaultScreen(st->dpy));
  st->gc = XCreateGC(st->dpy, st->window, 0, NULL);

  XGetWindowAttributes (st->dpy, st->window, &xgwa);
  ifs_reshape(st->dpy, st->window, st, xgwa.width, xgwa.height);
	
  st->pscale = 1;
  if (xgwa.width > 2560 || xgwa.height > 2560)
    st->pscale *= 3;  /* Retina displays */
  /* We aren't increasing the spacing between the pixels, just the size. */


  st->ncolours = get_integer_resource(st->dpy, "colors", "Colors");
  if (st->ncolours < st->lensnum)
    st->ncolours = st->lensnum;
  if (st->colours) free(st->colours);
  if (st->ncolours < 1) st->ncolours = 1;
  st->colours = (XColor *)calloc(st->ncolours, sizeof(XColor));
  if (!st->colours) exit(1);
  make_smooth_colormap (xgwa.screen, xgwa.visual, xgwa.colormap, 
                        st->colours, &st->ncolours,
                        True, 0, False);

  /* Initialize IFS data */
 
  st->delay = get_integer_resource(st->dpy, "delay", "Delay");
  st->length = get_integer_resource(st->dpy, "length", "Detail");
  if (st->length < 0) st->length = 0;
  st->mode = get_integer_resource(st->dpy, "mode", "Mode");

  st->rotate    = get_boolean_resource(st->dpy, "rotate", "Boolean");
                  /* Xft uses 'scale' */
  st->scale     = get_boolean_resource(st->dpy, "ifsScale", "Boolean");
  st->translate = get_boolean_resource(st->dpy, "translate", "Boolean");
  st->recurse = get_boolean_resource(st->dpy, "recurse", "Boolean");
  st->multi = get_boolean_resource(st->dpy, "multi", "Boolean");

  st->lensnum = get_integer_resource(st->dpy, "lensnum", "Functions");
  if (st->lenses) free (st->lenses);
  st->lenses = (Lens *)calloc(st->lensnum, sizeof(Lens));
  if (!st->lenses) exit(1);
  for (i = 0; i < st->lensnum; i++) {
    CreateLens(st,
	       myrandom(1)-0.5,
	       myrandom(1),
	       myrandom(4)-2,
	       myrandom(4)+2,
	       &st->lenses[i]);
  }

  return st;
}

static void
ifs_reshape (Display *dpy, Window window, void *closure, 
	     unsigned int w, unsigned int h)
{
  struct state *st = (struct state *)closure;
  XWindowAttributes xgwa;

  /* oh well, we need the screen depth anyway */
  XGetWindowAttributes (st->dpy, st->window, &xgwa);

  st->width = xgwa.width;
  st->widthb = ((xgwa.width + 31) >> 5);
  st->height = xgwa.height;
  st->width8 = xgwa.width << 8;
  st->height8 = xgwa.height << 8;

  if (!st->xmax && !st->ymax && !st->xmin && !st->ymin) {
    st->xmin = xgwa.width + 1;
    st->xmax = st->ymax = -1;
    st->ymin = xgwa.height + 1;
  }

  if (st->backbuffer != None && st->backbuffer != st->window) {
    XFreePixmap(st->dpy, st->backbuffer);
    st->backbuffer = None;
  }

  if (get_boolean_resource (st->dpy, "doubleBuffer", "Boolean")) {
    st->backbuffer = XCreatePixmap(st->dpy, st->window, st->width, st->height, xgwa.depth);
    XSetForeground(st->dpy, st->gc, st->blackColor);
    XFillRectangle(st->dpy, st->backbuffer, st->gc,
		   0, 0, st->width, st->height);
  } else {
    st->backbuffer = st->window;
    XClearWindow(st->dpy, st->window);
  }

  if (st->board) free(st->board);
  st->board = (unsigned int *)calloc(st->widthb * st->height, sizeof(unsigned int));
  if (!st->board) exit(1);
}

static Bool
ifs_event (Display *dpy, Window window, void *closure, XEvent *event)
{
  struct state *st = (struct state *)closure;
  if (screenhack_event_helper (dpy, window, event))
    {
      int i;
      for (i = 0; i < st->lensnum; i++) {
        CreateLens(st,
                   myrandom(1)-0.5,
                   myrandom(1),
                   myrandom(4)-2,
                   myrandom(4)+2,
                   &st->lenses[i]);
      }
      return True;
    }
  return False;
}

static void
ifs_free (Display *dpy, Window window, void *closure)
{
  struct state *st = (struct state *) closure;

  if (st->board) free(st->board);
  if (st->lenses) free(st->lenses);
  if (st->colours) free(st->colours);
  if (st->backbuffer != None && st->backbuffer != st->window)
    XFreePixmap(st->dpy, st->backbuffer);
  XFreeGC (dpy, st->gc);
  free(st);
}

XSCREENSAVER_MODULE ("IFS", ifs)
